### PR TITLE
[HttpKernel] Enhance MapRequestPayload adding format and validation group

### PR DESCRIPTION
**Pull request by renanbr**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Implements https://github.com/symfony/symfony/pull/49138#issuecomment-1500267966
| License       | MIT
| Doc PR        | not yet

This PR enhances `MapRequestPayload` (merged via https://github.com/symfony/symfony/pull/49138) by allowing users to restrict format and pick validation groups.

```php
# src/Controller/HelloController.php

namespace App\Controller;

use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\Attribute\AsController;
use Symfony\Component\HttpKernel\Attribute\MapRequestPayload;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\Validator\Constraints as Assert;

#[AsController]
final class HelloController
{
    #[Route('/hello', methods: ['POST'])]
    public function __invoke(
        #[MapRequestPayload(acceptFormat: 'json', validationGroups: 'strict')] MyObject $object,
    ): Response {
        // framework will...
        //     1. deserialized HTTP content from json, and throws exception in case of mismatch
        //     2. validate ONLY constraint in the "strict" group, and throws exception in case of invalidation
    }
}
```

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php [hunk 1]
```diff
@@ -137,12 +137,17 @@ private function mapRequestPayload(Request $request, string $type, MapRequestPay
             throw new HttpException(Response::HTTP_UNSUPPORTED_MEDIA_TYPE, 'Unsupported format.');
         }
 
+        $acceptedFormats = \is_string($attribute->format) ? [$attribute->format] : $attribute->format;
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: $attribute->acceptedFormats? to clarify it's not the actual format

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php [hunk 2]
```diff
@@ -114,19 +114,19 @@ public function resolve(Request $request, ArgumentMetadata $argument): iterable
...
         if (!$data = $request->query->all()) {
             return null;
         }
 
-        return $this->serializer->denormalize($data, $type, null, self::CONTEXT_DENORMALIZE + $attribute->context);
+        return $this->serializer->denormalize($data, $argument->getType(), null, self::CONTEXT_DENORMALIZE + $attribute->context);
     }
 
-    private function mapRequestPayload(Request $request, string $type, MapRequestPayload $attribute): ?object
+    private function mapRequestPayload(Request $request, ArgumentMetadata $argument, MapRequestPayload $attribute): ?object
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: seems irrelevant

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php [hunk 3]
```diff
@@ -22,7 +23,9 @@
 class MapRequestPayload extends ValueResolver
 {
     public function __construct(
+        public readonly array|string|null $format = null,
         public readonly array $context = [],
+        public readonly string|GroupSequence|array|null|false $groups = null,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: why `false`?

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php [hunk 4]
```diff
@@ -114,19 +114,19 @@ public function resolve(Request $request, ArgumentMetadata $argument): iterable
...
     {
         if ($data = $request->request->all()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: we should check the format before this if, so that forms are not accepted if only format "json" is listed for example

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php (hunk 3 above)
Comment (Koc) [reply src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: does it have sense to name it `validationGroups` in order to not mix with Serialization Groups?

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php (hunk 1 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: changed, thanks!

Changed existing `context` to `deserializationContext` as well to keep consistency.

Should we do the same in `MapQueryString`?

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php (hunk 2 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: The new `mapRequestPayload()` implementation requires `ArgumentMetadata` as an argument,
the call happens in a dynamic way (search for `$this->$payloadMapper()` in the `resolve()` method
so this change is required as the caller sends a new value

---

edit: I mean, its new implementation needs type **and** deserialization context

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php (hunk 3 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: @nicolas-grekas fixed, thanks for triggering it. It was a wrong copy&paste. In my local solution I'm able to disable groups defined from a preset, I removed it from this PR as it has no mean for now.

@Koc changed, thanks! Changed existing `context` to `$deserializationContext` as well to keep consistency. Should we do the same in `MapQueryString`?

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php (hunk 4 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: changed, thanks

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php [hunk 5]
```diff
@@ -22,7 +23,9 @@
 class MapRequestPayload extends ValueResolver
 {
     public function __construct(
-        public readonly array $context = [],
+        public readonly array|string|null $acceptedFormats = null,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: "acceptFormat"

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php [hunk 6]
```diff
@@ -22,7 +23,9 @@
...
+        public readonly array $deserializationContext = [],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: "serializationContext"

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php [hunk 7]
```diff
@@ -114,35 +114,40 @@ public function resolve(Request $request, ArgumentMetadata $argument): iterable
...
     }
 
-    private function mapRequestPayload(Request $request, string $type, MapRequestPayload $attribute): ?object
+    private function mapRequestPayload(Request $request, ArgumentMetadata $argument, MapRequestPayload $attribute): ?object
     {
+        if (null === $format = $request->getContentTypeFormat()) {
+            throw new HttpException(Response::HTTP_UNSUPPORTED_MEDIA_TYPE, 'Unsupported format.');
+        }
+
+        $acceptedFormats = \is_string($attribute->acceptedFormats) ? [$attribute->acceptedFormats] : $attribute->acceptedFormats;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: acceptFormat

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php [hunk 8]
```diff
@@ -114,35 +114,40 @@ public function resolve(Request $request, ArgumentMetadata $argument): iterable
...
+        if (\is_array($attribute->acceptedFormats) && !\in_array($format, $acceptedFormats, true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: `if ($attribute->acceptFormats && `

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: I also don't get this change: where do we use $argument and not only its type?

**Review by nicolas-grekas — COMMENTED** [review]
Almost GTM

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php (hunk 2 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php]: Fixed. Thanks for insisting on that, I wasn't seeing the problem.

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php (hunk 5 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: done

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php (hunk 6 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: done

File: src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php (hunk 5 above)
Comment (renanbr) [reply src/Symfony/Component/HttpKernel/Attribute/MapRequestPayload.php]: done

File: src/Symfony/Component/HttpKernel/Tests/Controller/ArgumentResolver/RequestPayloadValueResolverTest.php [hunk 9]
```diff
@@ -181,6 +181,14 @@ public function testRequestInputValidationPassed()
...
         $this->assertEquals($payload, $resolver->resolve($request, $argument)[0]);
     }
+
+    public function testNonSupportedRestrictedFormatMustThrowUnsupportedMediaTypeException()
+    {
+    }
+
+    public function testRestrictGroupValidation()
+    {
+    }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/Controller/ArgumentResolver/RequestPayloadValueResolverTest.php:372]: let's populate those now :)

File: src/Symfony/Component/HttpKernel/Tests/Controller/ArgumentResolver/RequestPayloadValueResolverTest.php [hunk 10]
```diff
@@ -181,10 +184,197 @@ public function testRequestInputValidationPassed()
...
+            $this->fail(sprintf('Expected "%s" to be thrown.', HttpException::class));
+        } catch (HttpException $e) {
+            $validationFailedException = $e->getPrevious();
+            $this->assertInstanceOf(ValidationFailedException::class, $validationFailedException);
+            $this->assertSame('title', $validationFailedException->getViolations()[0]->getPropertyPath());
+            $this->assertSame('This value is too short. It should have 10 characters or more.', $validationFailedException->getViolations()[0]->getMessage());
+        }
+    }
+
+    public function provideValidationGroupsOnManyTypes(): iterable
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/Controller/ArgumentResolver/RequestPayloadValueResolverTest.php]: all data providers should be static

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @renanbr.
